src/formatting_stack/util/diff.clj

```diff
@@ -0,0 +1,69 @@
+(ns formatting-stack.util.diff
+  (:require
+   [clojure.java.data :as data]
+   [clojure.java.io :as io]
+   [clojure.spec.alpha :as spec]
+   [clojure.string :as string]
+   [formatting-stack.util :refer [rcomp]]
+   [nedap.speced.def :as speced]
+   [nedap.utils.spec.predicates :refer [present-string?]])
+  (:import
+   (difflib DiffUtils)
+   (io.reflectoring.diffparser.api UnifiedDiffParser)
+   (java.io File)
+   (java.util.regex Pattern)))
+
+(spec/def ::start pos-int?)
+(spec/def ::end pos-int?)
+(spec/def ::filename present-string?)
+
+(speced/def-with-doc ::line-numbers
+  "maps of consecutive removals/changes (as `:start` to `:end`) per `:filename`."
+  (spec/coll-of (spec/keys :req-un [::start ::end ::filename])))
+
+(speced/defn ^::line-numbers diff->line-numbers
+  [^string? diff]
+  (->> (.parse (UnifiedDiffParser.) (io/input-stream (.getBytes diff)))
```

Comment:
Perhaps as a measure of caution, we can wrap this code with a try/catch.

In the catch clause, we return a `::line-numbers` vector with size 1; its object would have line 0 and column 0 but the actual filename parsed from the `diff` argument.

This way, linters always report something even if diff parsing went wrong.

Point being, f-s users benefit more from a less expressive linting report ("line 0 column 0") than from an extraneous stacktrace.

Of course, we can also print that something went wrong so that these aren't swallowed opaquely.

---

(especially if doing this, please add :min-count 1 to the ::line-numbers spec)

Reply:
> Perhaps as a measure of caution, we can wrap this code with a try/catch.

Seems like a good idea. However, multiple filenames can be present in 1 diff, and if parsing fails we'd still need some kind of backup parsing. 

Pulling the safe-guard one level higher, such as here: 
https://github.com/nedap/formatting-stack/blob/3360e801c4f34c8c4672bf857dcec274982546e6/src/formatting_stack/formatters/clean_ns.clj#L80

Except it's already wrapped in a try-catch by `process-in-parallel!`, which neatly reports a `:formatting-stack/report-processing-error`.

I think we should simply alpha this for a while to gain more confidence in this approach.

> (especially if doing this, please add :min-count 1 to the ::line-numbers spec)

A diff has no guarantee to contain any removed lines. an empty collection is a valid return value.

Reply:
> Pulling the safe-guard one level higher, such as here

Needless to say, it'd be better to have a single, abstract point for logic instead of repeating it around. The `UnifiedDiffParser` is the part that can plausibly fail, a try/catch should be as close to it as possible.

> I think we should simply alpha this for a while to gain more confidence in this approach.

As you can understand, an alpha cannot prove the lack of issues, particularly when we work in different teams. I'd be more comfortable with a try/catch; if you aren't, why don't we use a feature flag?

Example:

```clj
(defn diff->line-numbers* ...) ;; old impl, unaltered

(def diff->line-numbers
  (cond-> diff->line-numbers
    (System/getProperty "formatting-stack.diff.try") wrap-try))
```

Reply:
> A diff has no guarantee to contain any removed lines. an empty collection is a valid return value.

My reasoning was: if one is invoking `diff->line-numbers` in the first place, it's because some linter has something to report. But if the return value is `[]`, then nothing is being reported.

WDYT?

Reply:
My main concern with the try-catch is that it should live inside the linter in order to have enough context to create a meaningful fallback-result. `diff->line-numbers` should throw an exception on invalid diffs (might need to reflect it in the docstring, so consumers can expect it).

> if one is invoking diff->line-numbers in the first place, it's because some linter has something to report.

In our current usage: yes. I think this makes sense but be live one level higher. The linter should warn / error if a diff was generated, but no line-numbers could be determined.

Reply:
> I think this makes sense but be live one level higher.

Thanks, yes I suspected this was the case. Would be happy for any kind of approach

Reply:
`diff->line-numbers` is always used inside `process-in-parallel!`, so I added a test which verifies those exceptions are caught :).